### ISSUE TITLE
Kubernetes Applications do not autoconfigure bindings

### ISSUE THREAD

**Issue by derrick-solstice** [spring-cloud/spring-cloud-dataflow#3615]
**Problem description:**
While developing spring cloud stream applications to use with Kafka to be deployed on the Kubernetes platform it is not documented what docker configurations are required to support skipper configuration bindings.

**Solution description:**
Documentation should explain the need for dockerization of spring applications and the necessary configuration to support Skipper Bindings. 

However, this may be a deeper issue.  I have been under the impression that spring cloud stream automatically set the `spring.cloud.stream.bindings.input/output.destination` properties based on the stream and application name.  This does not appear to function with custom applications on Kubernetes, but does appear to work with the built-in applications.

**Description of alternatives:**
At current state, dockerized applications were built to define their own application bindings and the graphical and stream DSL were used largely for visual assistance to understand the stream.

**Additional context:**
If the issue is related to a docker image configuration change needed, please direct me to proper instructions and i'll open a PR to the doc site.  If this is a bug i can help reproduce.

**Comment by sabbyanandan** [comment]
Hi, @derrick-solstice. For custom Stream applications, if the channel bindings revolve around the default contracts in Spring Cloud Stream (e.g., `Source`, `Processor`, or `Sink`), then the corresponding input/output channel mapping is automatically applied by SCDF when it deploys the custom applications. 

If you implement a custom channel interface, however, SCDF will not be able to interpret it in this situation. We have a [Stream App DSL](https://dataflow.spring.io/docs/feature-guides/streams/stream-application-dsl/) to support deploying non-linear and comprehensive applications, including Kafka Streams processors.

As for,
> Documentation should explain the need for dockerization of spring applications and the necessary configuration to support Skipper Bindings.

It is not clear what do you mean by this. Perhaps you could elaborate on it with an example. Feel free to share a sample of the custom app, so we can relate to what you're describing.

**Comment by sabbyanandan** [comment]
Thanks, @derrick-solstice. If the custom app is simply leveraging the stock SCSt channel interfaces and that the default channel bindings are retained, there's no special handling is necessary. They should just work when deployed from SCDF on K8s. So yeah, please share the GH repo of the custom app, we can review.

As for Docker images for the OOTB apps, you can see [what we do in the plugin that generates the assembly.xml](https://github.com/spring-cloud/spring-cloud-app-starters-maven-plugins/blob/31d989ff4bb9ab8a4bac6b0f55b2cdf6a436161a/spring-cloud-stream-app-maven-plugin/src/main/java/org/springframework/cloud/stream/app/plugin/ScsProjectGenerator.java#L64) for each app-starter at the build time.

For comparison, when you run the `./mvnw -U clean install -PgenerateApps` on any of the app-starters repo, you will be able to find the container configurations, which will then be used for generating the docker images.

**Comment by derrick-solstice** [comment]
Afternoon @sabbyanandan,

I have generated a very simple [source application](https://github.com/derrick-solstice/scdf-demo-generator-kubernetes) that generates a message every second.  This application was then built into a docker container using the Dockerfile in that container.  

I then spun up a 3 node cluster in Google Kubernetes Engine.  Installed Spring Cloud Data flow using the current stable helm chart with kafka enabled.  I then imported the kafka docker applications and added my application from [docker hub](https://hub.docker.com/r/deanderson/demo-scdf-generator).

I deployed the stream with all default deployment parameters from SCDF.  I then monitored the logs, no messages appeared in the log application.

I have also attached the pod description below from my application.
```Name:               [[OMITTED]]
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               [[OMITTED]]
Start Time:         Tue, 05 Nov 2019 16:39:15 -0600
Labels:             pod-template-hash=56bbd48b45
                    role=spring-app
                    spring-app-id=test-demo-generator-v2
                    spring-deployment-id=test-demo-generator-v2
                    spring-group-id=test
Annotations:        <none>
Status:             Running
IP:                 10.48.1.10
Controlled By:      ReplicaSet/test-demo-generator-v2-56bbd48b45
Containers:
  test-demo-generator-v2:
    Container ID:  docker://[[OMITTED]]
    Image:         deanderson/demo-scdf-generator:1.0.0
    Image ID:      docker-pullable://deanderson/demo-scdf-generator[[OMITTED]]
    Port:          8080/TCP
    Host Port:     0/TCP
    Args:
      --spring.metrics.export.triggers.application.includes=integration**
      --spring.cloud.dataflow.stream.app.label=demo-generator
      --spring.cloud.stream.metrics.key=test.demo-generator.${spring.cloud.application.guid}
      --spring.cloud.stream.bindings.output.producer.requiredGroups=test
      --spring.cloud.stream.metrics.properties=spring.application.name,spring.application.index,spring.cloud.application.*,spring.cloud.dataflow.*
      --spring.cloud.dataflow.stream.name=test
      --spring.cloud.stream.bindings.output.destination=test.demo-generator
      --spring.cloud.dataflow.stream.app.type=source
    State:          Running
      Started:      Tue, 05 Nov 2019 16:39:18 -0600
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  1Gi
    Requests:
      cpu:      500m
      memory:   1Gi
    Liveness:   http-get http://:8080/actuator/health delay=90s timeout=2s period=60s #success=1 #failure=3
    Readiness:  http-get http://:8080/actuator/info delay=120s timeout=2s period=10s #success=1 #failure=3
    Environment:
      SPRING_CLOUD_STREAM_KAFKA_BINDER_BROKERS:   [[OMITTED]]
      SPRING_CLOUD_STREAM_KAFKA_BINDER_ZK_NODES:  [[OMITTED]]
      SPRING_CLOUD_APPLICATION_GUID:               (v1:metadata.uid)
      SPRING_CLOUD_APPLICATION_GROUP:             test
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from [[OMITTED]] (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  [[OMITTED]]:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  [[OMITTED]]
    Optional:    false
QoS Class:       Guaranteed
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age    From                                               Message
  ----    ------     ----   ----                                               -------
  Normal  Scheduled  8m13s  default-scheduler                                  Successfully assigned default[[OMITTED]]
  Normal  Pulling    8m12s  kubelet, [[OMITTED]]  pulling image "deanderson/demo-scdf-generator:1.0.0"
  Normal  Pulled     8m10s  kubelet, [[OMITTED]]  Successfully pulled image "deanderson/demo-scdf-generator:1.0.0"
  Normal  Created    8m10s  kubelet, [[OMITTED]]  Created container
  Normal  Started    8m10s  kubelet,[[OMITTED]]  Started container```

**Comment by derrick-solstice** [comment]
@sabbyanandan  I do see the arguments being passed to the container, however i'm completely unknowledgeable on how Kubernetes pods and docker images are meant to interact in this context.

I assume from this that the EntryPoint style or my Entry point command for docker are incorrect.

**Comment by derrick-solstice** [comment]
Morning @sabbyanandan,
Any insights into the issue above?  I have been trying to locate the skipper file that set's these deployments up and have been having issues finding them.  

Thanks again for the assistance everyone.

**Comment by sabbyanandan** [comment]
Thanks for the details. I don't see anything different from what we normally tend to use in the `Dockerfile`. Here's an [example](https://github.com/sabbyanandan/eda/blob/master/userproducer/Dockerfile).

I do see the reported behavior with your `source` on GKE, and I'm checking a few things to make sure everything is set up correctly. (for instance, the test-support dependency is in "test" scope, and among others).

**Comment by derrick-solstice** [comment]
The only other thing i've been mulling over is that I am utilizing Java 11.  I'm not sure if there is a different way to pass env variables when launching a Jar on 11.

**Comment by sabbyanandan** [comment]
FWIW, the same behavior is also noticed when you run your app against the `log-sink` locally, outside of SCDF even. See below.

> java -jar build/libs/demo-0.0.1-SNAPSHOT.jar --spring.cloud.stream.bindings.output.destination=t1

> java -jar log-sink-kafka-2.1.1.RELEASE.jar --spring.cloud.stream.bindings.input.destination=t1 --server.port=90011

I'm going to downgrade your app to the previous release to see what happens next.

**Comment by sabbyanandan** [comment]
I see it working against `Greenwich.SR2` release-train. The custom source is using the following dependencies.

```
± jar tvf build/libs/demo-0.0.1-SNAPSHOT.jar| grep stream
 59700 Wed Nov 06 11:10:50 PST 2019 BOOT-INF/lib/spring-cloud-stream-binder-kafka-2.1.3.RELEASE.jar
 37428 Wed Nov 06 11:10:50 PST 2019 BOOT-INF/lib/spring-cloud-stream-binder-kafka-core-2.1.3.RELEASE.jar
293597 Wed Nov 06 11:10:50 PST 2019 BOOT-INF/lib/spring-cloud-stream-2.1.3.RELEASE.jar
 11369 Wed Nov 06 10:37:22 PST 2019 BOOT-INF/lib/reactive-streams-1.0.3.jar
```

@sobychacko is digging as to why it is not working on Horsham.RC1.

**Comment by derrick-solstice** [comment]
@sabbyanandan 
I was utilizing `Hoxton.RC1` as the Spring Cloud Version.  It's good to know that this isn't a facet of my own making.  My demo applications were on `Hoxton.M3` as the SCV.

To confirm the workaround currently is to downgrade to the `Greenwich.SR2`  Spring Cloud Version for now? I'll await bug confirmation first.

**Comment by sabbyanandan** [comment]
@derrick-solstice: There's a side effect in Spring Kafka that we are troubleshooting (see: https://github.com/spring-projects/spring-kafka/issues/1301). Depending on whatever comes out of it, we may have to adjust things in SCSt Kafka Binder, as well.

For now, yes, you can switch to Greenwich + Germantown dependencies.

**Comment by sabbyanandan** [comment]
The associated Kafka Binder issue: https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/796.

**Comment by derrick-solstice** [comment]
@sabbyanandan ,

Thank you for the clarification!  I'll track the spring-kafka issue for the long term resolution and for my demo code I'll revert to Spring Cloud Version `Greenwich.SR2`.

I'll confirm tomorrow if this resolves the issue.  You all rule! Great job isolating this issue.

**Comment by derrick-solstice** [comment]
@sabbyanandan 

I have not yet been able to verify this. I will do so by monday at the latest. Thank you again.

**Comment by sobychacko** [comment]
@derrick-solstice @sabbyanandan There was an interoperability issue we discovered between Spring Cloud Stream 2.x and 3.0 apps. This is addressed since then as part of this [issue](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/796). Could you use `Horsham.RC2` (through spring-cloud `Hoxon.RC2`) and verify?

**Comment by sabbyanandan** [comment]
I am closing this since it had been fixed and released in SCSt 3.0 RC2.

**Comment by derrick-solstice** [comment]
Thank you for verifiying,  I was able to redeploy and verify this yesterday.  

You guys are awesome!!
